ena-submission/scripts/create_project.py

```diff
@@ -0,0 +1,380 @@
+    )
+    logger.debug(
+        (
+            f"Found {len(submitting_project)} entries in submission_table in",
+            " status SUBMITTING_PROJECT",
+        )
+    )
+    for row in submitting_project:
+        group_key = {"group_id": row["group_id"], "organism": row["organism"]}
+        seq_key = {"accession": row["accession"], "version": row["version"]}
+
+        # 1. check if there exists an entry in the project table for (group_id, organism)
+        corresponding_project = find_conditions_in_db(
+            db_config, table_name="project_table", conditions=group_key
+        )
+        if len(corresponding_project) == 1 and corresponding_project[0]["status"] == str(
+            Status.SUBMITTED
+        ):
+            update_values = {
+                "status_all": StatusAll.SUBMITTED_PROJECT,
+                "center_name": corresponding_project[0]["center_name"],
+            }
+            update_db_where_conditions(
+                db_config,
+                table_name="submission_table",
+                conditions=seq_key,
+                update_values=update_values,
+            )
+        if len(corresponding_project) == 0:
+            error_msg = (
+                "Entry in submission_table in status SUBMITTING_PROJECT",
+                " with no corresponding project",
+            )
+            raise RuntimeError(error_msg)
+
+
+def project_table_create(db_config, config, retry_number=3):
+    """
+    1. Find all entries in project_table in state READY
+    2. Create project_set: get_group_info from loculus, use entry and config for other fields
+    3. Update project_table to state SUBMITTING (only proceed if update succeeds)
+    4. If (create_ena_project succeeds): update state to SUBMITTED with results
+    3. Else update state to HAS_ERRORS with error messages
+    """
+    ena_config = get_ena_config(
+        config.ena_submission_username,
+        config.ena_submission_password,
+        config.ena_submission_url,
+        config.ena_reports_service_url,
+    )
+    conditions = {"status": Status.READY}
+    ready_to_submit_project = find_conditions_in_db(
+        db_config, table_name="project_table", conditions=conditions
+    )
+    logger.debug(f"Found {len(ready_to_submit_project)} entries in project_table in status READY")
+    for row in ready_to_submit_project:
+        group_key = {"group_id": row["group_id"], "organism": row["organism"]}
+
+        try:
+            group_info = get_group_info(config, row["group_id"])[0]["group"]
+        except Exception as e:
+            logger.error(f"Was unable to get group info for group: {row["group_id"]}, {e}")
+            continue
+
+        project_set = construct_project_set_object(group_info, config, row, test=True)
+        update_values = {
+            "status": Status.SUBMITTING,
+            "started_at": datetime.now(tz=pytz.utc),
+            "center_name": group_info["institution"],
```

Comment:
Is a SQL injection possible through the `institutions` field? Because it is passed on into `update_db_where_conditions` which uses string concatenation to construct the SQL.

Reply:
I have modified the function to now always use `%s` for passing inputs and validating the name.